File: app/src/main/java/com/adiaz/deportesmadrid/adapters/FavoritesAdapter.java
package com.adiaz.deportesmadrid.adapters;

import android.content.Context;
import android.support.annotation.Nullable;
import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.adiaz.deportesmadrid.R;
import com.adiaz.deportesmadrid.db.daos.GroupsDAO;
import com.adiaz.deportesmadrid.db.entities.Group;
import com.adiaz.deportesmadrid.db.entities.Favorite;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;

public class FavoritesAdapter extends RecyclerView.Adapter<FavoritesAdapter.ViewHolder> {

    Context mContext;
    FavoritesAdapter.ListItemClickListener mListItemClickListener;
    List<Favorite> mFavoritesList;

    private static final Integer TYPE_TEAM = 1;
    private static final Integer TYPE_GROUP = 2;

    public FavoritesAdapter(Context context, ListItemClickListener listItemClickListener, List<Favorite> mFavoritesList) {
        this.mContext = context;
        this.mFavoritesList = mFavoritesList;
        this.mListItemClickListener = listItemClickListener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        int layoutId = R.layout.listitem_favorites_team;
        if (viewType==TYPE_GROUP) {
            layoutId = R.layout.listitem_favorites_group;
        }
        LayoutInflater layoutInflater = LayoutInflater.from(mContext);
        View view = layoutInflater.inflate(layoutId, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        if (mFavoritesList!=null) {
            Group group = GroupsDAO.queryCompetitionsById(mContext, mFavoritesList.get(position).idGroup());
            holder.tvGroup.setText(group.nomGrupo());
            holder.tvFase.setText(group.nomFase());
            holder.tvCompetition.setText(group.nomCompeticion());
            String teamName = mFavoritesList.get(position).idTeam();
            if (teamName!=null) {
                holder.tvTeam.setText(teamName);
            }
        }
    }

    @Override
    public int getItemCount() {
        return mFavoritesList.size();
    }

    @Override
    public int getItemViewType(int position) {
        if (mFavoritesList.get(position).idTeam()==null) {
            return TYPE_GROUP;
        } else {
            return TYPE_TEAM;
        }
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        @BindView(R.id.cv_favorites)
        CardView cvFavorites;

        @Nullable
        @BindView(R.id.tv_team)
        TextView tvTeam;

        @Nullable
        @BindView(R.id.tv_group)
        TextView tvGroup;

        @Nullable
        @BindView(R.id.tv_fase)
        TextView tvFase;

        @Nullable
        @BindView(R.id.tv_competition)
        TextView tvCompetition;

        ViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
            cvFavorites.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            mListItemClickListener.onListItemClick(getAdapterPosition());
        }
    }

    public interface ListItemClickListener {
        void onListItemClick(int clickedItemIndex);
    }
}

File: app/src/main/java/com/adiaz/deportesmadrid/adapters/FavoritesAdapter.kt
package com.adiaz.deportesmadrid.adapters

import android.content.Context
import android.support.v7.widget.CardView
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.widget.TextView

import com.adiaz.deportesmadrid.R
import com.adiaz.deportesmadrid.db.daos.GroupsDAO
import com.adiaz.deportesmadrid.db.entities.Favorite
import com.adiaz.deportesmadrid.extensions.layoutInflater
import com.adiaz.deportesmadrid.utils.Constants

import kotlinx.android.synthetic.main.listitem_favorites_team.view.*

class FavoritesAdapter(internal var mContext: Context, listItemClickListener: ListItemClickListener, internal var mFavoritesList: List<Favorite>?) : RecyclerView.Adapter<FavoritesAdapter.ViewHolder>() {

    private var mListItemClickListener: FavoritesAdapter.ListItemClickListener = listItemClickListener

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        var layoutId = R.layout.listitem_favorites_team
        if (viewType == TYPE_GROUP) {
            layoutId = R.layout.listitem_favorites_group
        }
        val view = mContext.layoutInflater.inflate(layoutId, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        if (mFavoritesList != null && mFavoritesList!!.size >= position) {
            val group = GroupsDAO.queryCompetitionsById(mContext, mFavoritesList!![position].idGroup)
            if (group != null) {
                holder.tvGroup.text = group.nomGrupo
                holder.tvFase.text = group.nomFase
                holder.tvCompetition.text = group.nomCompeticion
                holder.tvPath.text = group.deporte + Constants.PATH_SEPARATOR + group.distrito + Constants.PATH_SEPARATOR + group.categoria
                val teamName = mFavoritesList!![position].nameTeam
                if (teamName != null) {
                    holder.tvTeam!!.text = teamName
                }
            }
        }
    }

    override fun getItemCount(): Int {
        return mFavoritesList!!.size
    }

    override fun getItemViewType(position: Int): Int {
        return if (mFavoritesList!![position].idTeam == null) {
            TYPE_GROUP
        } else {
            TYPE_TEAM
        }
    }

    inner class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {

        internal var cvFavorites: CardView = itemView.cardView
        internal var tvTeam: TextView? = itemView.tvTeam
        internal var tvGroup: TextView = itemView.tvGroup
        internal var tvFase: TextView = itemView.tvFase
        internal var tvCompetition: TextView = itemView.tvCompetition
        internal var tvPath: TextView = itemView.tvPath

        init {
            cvFavorites.setOnClickListener(this)
        }

        override fun onClick(view: View) {
            mListItemClickListener.onListItemClick(adapterPosition)
        }
    }

    interface ListItemClickListener {
        fun onListItemClick(clickedItemIndex: Int)
    }

    companion object {
        private const val TYPE_TEAM = 1
        private const val TYPE_GROUP = 2
    }
}
